File: 25-Dynamic-Programming/java/01-MaxProductSubarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int result = nums[0];
        int prefix = 0;
        int suffix = 0;
        int n = nums.length;

        for (int i = 0; i < n; i++) {
            prefix = (prefix == 0 ? 1 : prefix) * nums[i];
            suffix = (suffix == 0 ? 1 : suffix) * nums[n - 1 - i];
            result = Math.max(result, Math.max(prefix, suffix));
        }

        return result;
    }
}

File: 25-Dynamic-Programming/python/01_max_product_subarray.py
from typing import List

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        result = nums[0]
        prefix = 0
        suffix = 0
        n = len(nums)

        for i in range(n):
            prefix = (1 if prefix == 0 else prefix) * nums[i]
            suffix = (1 if suffix == 0 else suffix) * nums[n - 1 - i]
            result = max(result, prefix, suffix)

        return result
